python-checks/src/main/java/org/sonar/python/checks/PandasChainInstructionCheck.java

```diff
@@ -0,0 +1,136 @@
+  }
+
+
+  private void checkChainedInstructions(SubscriptionContext ctx) {
+    QualifiedExpression expr = (QualifiedExpression) ctx.syntaxNode();
+    List<QualifiedExpression> chain = visitParent(expr, new ArrayList<>());
+    if (chain.size() >= MAX_CHAIN_LENGTH && isValidPandasCall(chain)) {
+      ctx.addIssue(chain.iterator().next(), MESSAGE);
+    }
+  }
+
+  private List<QualifiedExpression> visitParent(QualifiedExpression expr, List<QualifiedExpression> chain) {
+    if (visited.contains(expr)) {
+      return chain;
+    }
+    visited.add(expr);
+    chain.add(expr);
+
+    if (expr.qualifier().is(Tree.Kind.CALL_EXPR)) {
+      return visitCalleeParent((CallExpression) expr.qualifier(), chain);
+    } else if (expr.qualifier().is(Tree.Kind.SUBSCRIPTION)) {
+      return ignoreSubscriptionAndGetCallExpr((SubscriptionExpression) expr.qualifier())
+        .map(callExpr -> visitCalleeParent(callExpr, chain))
+        .orElse(chain);
+    }
+    return chain;
+  }
+
+
+  private List<QualifiedExpression> visitCalleeParent(CallExpression call, List<QualifiedExpression> chain) {
+    return Optional.of(call.callee())
+      .flatMap(TreeUtils.toOptionalInstanceOfMapper(QualifiedExpression.class))
+      .map(qe -> visitParent(qe, chain))
+      .orElse(chain);
+  }
+
+  private static Optional<CallExpression> ignoreSubscriptionAndGetCallExpr(SubscriptionExpression qualifier) {
+    if (qualifier.object().is(Tree.Kind.CALL_EXPR)) {
+      return Optional.of((CallExpression) qualifier.object());
+    } else if (qualifier.object().is(Tree.Kind.SUBSCRIPTION)) {
+      return ignoreSubscriptionAndGetCallExpr((SubscriptionExpression) qualifier.object());
+    }
+    return Optional.empty();
+  }
+
+  private static boolean isValidPandasCall(List<QualifiedExpression> chain) {
+    QualifiedExpression firstQualifiedExpression = chain.get(chain.size() - 1);
+
+    boolean isADataFrameMethodCall = Optional.ofNullable(firstQualifiedExpression.symbol())
+      .map(Symbol::fullyQualifiedName)
+      .filter(fqn -> fqn.startsWith(DATAFRAME_FQN))
+      .isPresent();
+
+    boolean isAFunctionReturningADataFrame = Optional.ofNullable(firstQualifiedExpression.symbol())
+      .flatMap(PandasChainInstructionCheck::isReturnTypeADataFrame)
+      .orElse(false);
+
+    boolean doesNotContainACallToPipe = chain.stream()
+      .map(QualifiedExpression::symbol)
+      .filter(Objects::nonNull)
+      .map(Symbol::fullyQualifiedName)
+      .filter(Objects::nonNull)
+      .noneMatch((DATAFRAME_FQN + ".pipe")::equals);
+
+    boolean isADataFrame = "DataFrame".equals(InferredTypes.typeName(firstQualifiedExpression.qualifier().type()));
+
+    return (isADataFrameMethodCall || isAFunctionReturningADataFrame || isADataFrame) && doesNotContainACallToPipe;
+  }
+
+  private static Optional<Boolean> isReturnTypeADataFrame(Symbol symbol) {
+    return Optional.of(symbol)
+      .filter(s -> s.is(Symbol.Kind.AMBIGUOUS))
```

Comment:
What if the symbol is a `FUNCTION`? I understand most Pandas methods are ambiguous so in practice, this should work, but would we then have FNs if a method only has one definition?

Reply:
Yes, I think I could add the happy path here :).